### PR TITLE
BUG: Support big-endian platforms and big-endian WAVs

### PR DESCRIPTION
**Pull request by nforro**
PR #12287 added support for reading arbitrary-bit-depth WAVs, but the code doesn't consider big-endian WAVs, and doesn't work as expected on big-endian platforms due to the use of native-byte-order data-types. This change fixes that.

There is also a simple test case that compares euqivalent RIFX (big-endian) and RIFF (little-endian) files to verify the data read are the same.

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
This looks really promising at first glance--maybe @endolith can look it over

**Comment by endolith** [comment]
I don't know of anything that can open `RIFX` files, but `test-44100Hz-be-1ch-4bytes.wav` doesn't look correct:

image

this would be a RIFF chunk size of 0x52 = 82 bytes?  `test-44100Hz-le-1ch-4bytes.wav` has:

image

which is a RIFF chunk size of 0x4524 = 17700 bytes, while the actual file is 17720 bytes long.

Oh it has the same length as `test-8000Hz-be-3ch-5S-24bit.wav`, which is correct:

image


Actually I think `test-44100Hz-le-1ch-4bytes.wav` is wrong, too, since it claims 0x4524 = 17700, but the number of bytes after that is 0x4530 = 17712 bytes.  Riffpad seems to say so too:

image

I guess we only pay attention to the RIFF chunk size (`file_size`) while reading the `chunk_id`s, and then use the data subchunk's own size (`size`) to determine the output data length, even if it's larger than the RIFF chunk size?

**Comment by nforro** [comment]
Sorry, my bad, `test-44100Hz-be-1ch-4bytes.wav` should be fixed now.

File: scipy/io/wavfile.py [hunk 1]
```diff
@@ -458,10 +458,13 @@ def _read_data_chunk(fid, format_tag, channels, bit_depth, is_big_endian,
 
         if dtype == 'V1':
             # Rearrange raw bytes into smallest compatible numpy dtype
-            dt = numpy.int32 if bytes_per_sample == 3 else numpy.int64
-            a = numpy.zeros((len(data) // bytes_per_sample, dt().itemsize),
+            dt = f'{fmt}i4' if bytes_per_sample == 3 else f'{fmt}i8'
```
Comment (endolith) [inline scipy/io/wavfile.py:461]: This changes `dt` from type to string, but that should be fine in the two places it's used

**Review by endolith — APPROVED** [review]
Ok, the WAV files look good now. I noticed you added Front Center channel to the BE version, not sure if we should change that on older ones, too.

There's an "euqivalent" typo in the commit message.

Otherwise it looks good to me.

What do you use these big-endian files for, by the way?

**Comment by nforro** [comment]
> Ok, the WAV files look good now. I noticed you added Front Center channel to the BE version, not sure if we should change that on older ones, too.

Unintentionally, I generated the file with sox, but I suppose it makes sense, and I think it makes sense to fix it in the other files too, along with the weird RIFF chunk size. I'm not sure that's in scope of this pull request though.
 
> There's an "euqivalent" typo in the commit message.

Thanks, fixed.

> What do you use these big-endian files for, by the way?

I actually don't, but the old code supports big-endian WAVs, so I thought it would be nice if the new code supported them too.

**Review by larsoner — APPROVED** [review]
@endolith is happy and this LGTM so I'll go ahead and merge, thanks @nforro !

**Pull request merged**